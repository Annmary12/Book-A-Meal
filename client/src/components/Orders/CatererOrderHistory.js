import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import isEmpty from 'lodash/isEmpty';
import NoOrder from './NoOrder';
import styles from './order.scss';
import Table from '../ResponsiveTable/Table.js';
import Cell from '../ResponsiveTable/Cell.js';
import Row from '../ResponsiveTable/Row.js';
import CatererOrderHistoryRow from './CatererOrderHistoryRow';

/**
 * @class CatererOrderHistory
 */
class CatererOrderHistory extends Component {
  static propTypes = {
    fetchOrderHistory: PropTypes.func.isRequired,
    paginationDetails: PropTypes.objectOf(PropTypes.number).isRequired,
    deliverOrder: PropTypes.func.isRequired,
    orders: PropTypes.arrayOf(PropTypes.object).isRequired
  }

  /**
   * @returns {undefined} undefined
   */
  componentDidMount() {
    this.props.fetchOrderHistory();
  }

  handlePageClick = (data) => {
    this.props.fetchOrderHistory({
      page: data.selected + 1,
      limit: 10
    });
  }

  handleDeliver = (orderId) => {
    this.props.deliverOrder(orderId);
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    const { orders, deliverOrder, paginationDetails } = this.props;
    return (
      <div className={styles.container}>
        <h2>Order History</h2>
        {
          isEmpty(orders) ?
            <NoOrder /> :
            <div>
              <Table>
                <Row header>
                  <Cell>Date</Cell>
                  <Cell>Meal</Cell>
                  <Cell>Qty</Cell>
                  <Cell>Price</Cell>
                  <Cell>Delivery address</Cell>
                  <Cell>Status</Cell>
                  <Cell>Actions</Cell>
                </Row>
                {
                  orders.map(order => (
                    <CatererOrderHistoryRow
                      key={order.id}
                      order={order}
                      handleDeliver={deliverOrder}
                    />
                  ))
                }
              </Table>
              <ReactPaginate
                breakLabel={<a href="">...</a>}
                pageCount={paginationDetails.pageCount}
                marginPagesDisplayed={2}
                pageRangeDisplayed={5}
                previousClassName="previous"
                nextClassName="next"
                previousLabel="<<"
                nextLabel=">>"
                onPageChange={this.handlePageClick}
                containerClassName="pagination"
                subContainerClassName="pages pagination"
                activeClassName="active"
              />
              <div className={styles.paginationDetails}>
                {`Page ${paginationDetails.currentPage} of ${paginationDetails.pageCount}`}
              </div>
            </div>
          }
      </div>
    );
  }
}

export default CatererOrderHistory;
