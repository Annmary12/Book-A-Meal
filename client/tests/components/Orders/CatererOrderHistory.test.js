import React from 'react';
import CatererOrderHistory from '../../../src/components/Orders/CatererOrderHistory';

const props = {
  fetchOrderHistory: jest.fn(),
  paginationDetails: {
    currentPage: 1,
    pageCount: 1
  },
  deliverOrder: jest.fn(),
  orders: []
}
describe('CatererOrderHistory component', () => {
  it('should match snapshot if orders array is empty', () => {
    const wrapper = shallow(
      <CatererOrderHistory {...props} />
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot if orders array is not empty', () => {
    const wrapper = shallow(
      <CatererOrderHistory
        {...props}
        orders={[
          {
            id: 'asdf23e',
          }
        ]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchOrderHistory prop on mount', () => {
    const wrapper = shallow(
      <CatererOrderHistory {...props} />
    );

    wrapper.instance().componentDidMount();

    expect(props.fetchOrderHistory).toBeCalled();
  });

  describe('handlePageClick method', () => {
    it('should call fetchOrderHistory prop function with pagination details', () => {
      const wrapper = shallow(
        <CatererOrderHistory {...props} />
      );

      const data = {
        selected: 1
      }

      wrapper.instance().handlePageClick(data);

      expect(props.fetchOrderHistory).toBeCalledWith({
        page: data.selected + 1,
        limit: 10
      });
    });
  });

  describe('handleDeliver method', () => {
    it('should call deliver order prop function', () => {
      const wrapper = shallow(
        <CatererOrderHistory {...props} />
      );

      wrapper.instance().handleDeliver('fad');

      expect(props.deliverOrder).toBeCalledWith('fad');
    });
  });
});