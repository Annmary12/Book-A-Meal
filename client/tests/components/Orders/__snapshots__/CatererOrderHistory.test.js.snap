// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatererOrderHistory component should match snapshot if orders array is empty 1`] = `
<div
  className="container"
>
  <h2>
    Order History
  </h2>
  <NoOrder />
</div>
`;

exports[`CatererOrderHistory component should match snapshot if orders array is not empty 1`] = `
<div
  className="container"
>
  <h2>
    Order History
  </h2>
  <div>
    <Table>
      <Row
        header={true}
      >
        <Cell
          title=""
        >
          Date
        </Cell>
        <Cell
          title=""
        >
          Meal
        </Cell>
        <Cell
          title=""
        >
          Qty
        </Cell>
        <Cell
          title=""
        >
          Price
        </Cell>
        <Cell
          title=""
        >
          Delivery address
        </Cell>
        <Cell
          title=""
        >
          Status
        </Cell>
        <Cell
          title=""
        >
          Actions
        </Cell>
      </Row>
      <CatererOrderHistoryRow
        handleDeliver={[MockFunction]}
        key="asdf23e"
        order={
          Object {
            "id": "asdf23e",
          }
        }
      />
    </Table>
    <PaginationBoxView
      activeClassName="active"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel=">>"
      onPageChange={[Function]}
      pageCount={1}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel="<<"
      subContainerClassName="pages pagination"
    />
    <div
      className="paginationDetails"
    >
      Page 1 of 1
    </div>
  </div>
</div>
`;
